raphtory/src/serialise/mod.rs

```diff
@@ -6,5 +9,137 @@ mod proto {
     include!(concat!(env!("OUT_DIR"), "/serialise.rs"));
 }
 
+use std::{
+    fs::{self, File, OpenOptions},
+    io::{self, Read, Write},
+    path::{Path, PathBuf},
+};
+
 pub use proto::Graph as ProtoGraph;
 pub use serialise::{CacheOps, StableDecode, StableEncode};
+
+#[derive(Clone, Debug)]
+pub struct GraphFolder {
+    path: PathBuf,
+    prefer_zip_format: bool,
+}
+
+pub enum GraphReader {
+    Zip(Vec<u8>),
+    Folder(Mmap),
+}
+
+impl AsRef<[u8]> for GraphReader {
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            Self::Zip(bytes) => bytes.as_ref(),
+            Self::Folder(mmap) => mmap.as_ref(),
+        }
+    }
+}
+
+impl GraphFolder {
+    pub fn new_as_zip(path: impl AsRef<Path>) -> Self {
+        let folder: GraphFolder = path.into();
+        Self {
+            prefer_zip_format: true,
+            ..folder
+        }
+    }
+
+    // TODO: make it private again once we stop using it from the graphql crate
+    pub fn get_graph_path(&self) -> PathBuf {
+        self.path.join("graph")
+    }
+
+    pub fn read_graph(&self) -> Result<GraphReader, io::Error> {
+        if self.path.is_file() {
+            let file = File::open(&self.path)?;
+            let mut archive = ZipArchive::new(file)?;
+            let mut entry = archive.by_name("graph")?;
+            let mut buf = vec![];
+            entry.read_to_end(&mut buf)?;
+            Ok(GraphReader::Zip(buf))
+        } else {
+            let file = File::open(self.get_graph_path())?;
+            let buf = unsafe { memmap2::MmapOptions::new().map(&file)? };
+            Ok(GraphReader::Folder(buf))
+        }
+    }
+
+    pub fn write_graph(&self, buf: &[u8]) -> Result<(), io::Error> {
+        if self.prefer_zip_format {
+            let file = File::create(&self.path)?;
+            let mut zip = ZipWriter::new(file);
+            zip.start_file::<_, ()>("graph", FileOptions::default())?;
+            zip.write_all(buf)
+        } else {
```

Comment:
if the directory is empty is fine for writing a graph. For vectors we don't check